src/cli/README_HISTORY.md

```diff
@@ -0,0 +1,290 @@
+# OpenThread CLI - History Tracker
+
+History Tracker module records history of different events (e.g., RX and TX IPv6 messages or network info changes, etc.) as the Thread network operates. All tracked entries are timestamped.
+
+All commands under `history` require `OPENTHREAD_CONFIG_HISTORY_TRACKER_ENABLE` feature to be enabled.
+
+The number of entries recorded for each history list is configurable through a set of OpenThread config options, e.g. `OPENTHREAD_CONFIG_HISTORY_TRACKER_NET_INFO_LIST_SIZE` specifies the number of entries in Network Info history list. The History Tracker will keep the most recent entries overwriting oldest one when the list gets full.
+
+## Command List
+
+Usage : `history [command] ...`
+
+- [help](#help)
+- [netinfo](#netinfo)
+- [rx](#rx)
+- [tx](#tx)
+
+## Timestamp Format
+
+Recorded entries are timestamped. When the history list is printed, the timestamps are shown relative the time the command was issues (i.e., when the list was printed) indicating how long ago the entry was recorded.
+
+```bash
+> history netinfo
+| Time                     | Role     | Mode | RLOC16 | Partition ID |
++--------------------------+----------+------+--------+--------------+
+|         02:31:50.628 ago | leader   | rdn  | 0x2000 |    151029327 |
+|         02:31:53.262 ago | detached | rdn  | 0xfffe |            0 |
+|         02:31:54.663 ago | detached | rdn  | 0x2000 |            0 |
+Done
+```
+
+For example `02:31:50.628 ago` indicates the event was recorded "2 hours, 31 minutes, 50 seconds, and 628 milliseconds ago". Number of days is added for events that are older than 24 hours, e.g., `1 day 11:25:31.179 ago`, or `31 days 03:00:23.931 ago`.
+
+Timestamps use millisecond accuracy and are tacked up to 49 days. If the event is older than 49 days, the entry is still tracked in the list but the timestamp is shown as `more than 49 days ago`.
+
+## Command Details
+
+### help
+
+Usage: `history help`
+
+Print SRP client help menu.
+
+```bash
+> history help
+help
+netinfo
+rx
+tx
+Done
+>
+```
+
+### netinfo
+
+Usage `history netinfo [list] [<num-entries>]`
+
+Print the Network Info history. Each Network Info provides
+
+- Device Role
+- MLE Link Mode
+- RLOC16
+- Partition ID
+
+Print the Network Info history as a table.
+
+```bash
+> history netinfo
+| Time                     | Role     | Mode | RLOC16 | Partition ID |
++--------------------------+----------+------+--------+--------------+
+|         00:00:10.069 ago | router   | rdn  | 0x6000 |    151029327 |
+|         00:02:09.337 ago | child    | rdn  | 0x2001 |    151029327 |
+|         00:02:09.338 ago | child    | rdn  | 0x2001 |    151029327 |
+|         00:07:40.806 ago | child    | -    | 0x2001 |    151029327 |
```

Comment:
In what scenario do we get three `child` entries in a row?

Reply:
I did notice the same at the time. 😄 

On this node, I did the following, `mode -` , then attached to the network and then later issued `mode rdn`.

This line `00:07:40.806 ago | child    | -    | 0x2001 |    151029327` is for when role change to child (device attached first time)

The other two are for when "mode" changed. Mode change actually can be recorded as if two changes happened.
```
|         00:02:09.337 ago | child    | rdn  | 0x2001 |    151029327 |
|         00:02:09.338 ago | child    | rdn  | 0x2001 |    151029327 |
```

 The reason for this is tied to fact that mode change triggers once from `SetDeviceMode()` itself calling `HistoryTracker::RecordNetworkInfo()` (we don't have a `Notifier` event for mode change) and then we may also get another `Notifier` event (later) for RLOC added/removed events. 

Here is the call chain: `Mle::SetDeviceMode()` to set the new mode which can call `SetStateChild(GetRloc16())` which itself calls `SetRloc16(aRloc16)` which alwys removes and re-adds `mMeshLocal16` (even if the rloc16 is same), which trigger notififer event` kEventThreadRlocRemoved` and `kEventThreadRlocAdded`

I opted to let it behave like this for now, but was considering to address it later (in a seperate PR):
- One  solution is to add a `Notifier` event for mode change then all events are emitted together and we process it once. I slightly prefer this one. The only downside is consuming one of the bits in `OT_CHANGED_*` - we are already using 30 bits and it's 32 bit value so 2 bit left)
- We can also consider changing the `SetRlco16()` code to skip remove and re-add (if the `mMeshLocal16` happens to be same, but I am not sure if it may have other side-effects).

Thoughts/Ideas?

Reply:
Good points. I'm not sure if there's a good reason for always executing the code in `SetRloc16()` even if the short address does not change. I would prefer avoiding repeated execution if there is no good reason for it. I'm find with addressing this in a separate PR.

Reply:
I tried checking different code-paths. I think changing `SetRloc16()` to avoid removing and re-adding the `mMeshLocag16` address if remains unchanged should work fine. I will submit a PR to do this.